src/main/scala/sigmastate/serialization/ErgoTreeSerializer.scala

```diff
@@ -1,42 +1,79 @@
 package sigmastate.serialization
 
-import sigmastate.SType
-import sigmastate.Values.{Constant, ErgoTree, Value}
+import org.ergoplatform.ValidationRules.CheckDeserializedScriptIsSigmaProp
+import org.ergoplatform.{ValidationSettings, ValidationRules, ValidationRule, SoftForkException}
+import sigmastate.{SType, TrivialProp}
+import sigmastate.Values.{Value, ErgoTree, Constant}
 import sigmastate.lang.DeserializationSigmaBuilder
 import sigmastate.lang.Terms.ValueOps
 import sigmastate.lang.exceptions.SerializerException
 import sigmastate.utils.{SigmaByteReader, SigmaByteWriter}
+import sigma.util.Extensions._
+import sigmastate.Values.ErgoTree.EmptyConstants
 
 import scala.collection.mutable
 
 class ErgoTreeSerializer {
 
-  /** Default serialization of ErgoTree. Doesn't apply any transformations and guarantee to preserve original
-    * structure after deserialization. */
-  def serializeErgoTree(ergoTree: ErgoTree): Array[Byte] = {
+  private def serializeWithoutSize(ergoTree: ErgoTree): Array[Byte] = {
     val w = SigmaSerializer.startWriter()
     serializeHeader(ergoTree, w)
     ValueSerializer.serialize(ergoTree.root, w)
     w.toBytes
   }
 
+  /** Default serialization of ErgoTree. Doesn't apply any transformations and guarantee to preserve original
+    * structure after deserialization. */
+  def serializeErgoTree(ergoTree: ErgoTree): Array[Byte] = {
+    val bytes = serializeWithoutSize(ergoTree)
+    if (ergoTree.hasSize) {
```

Comment:
What if `ergoTree.hasSize = false` and also there is an op-code from soft-fork? Looks like new nodes will accept the script, while the old one won't and will reject the block.